ocaml/xapi/repository.ml

```diff
@@ -670,6 +673,63 @@ let update_cache ~host ~failed_livepatches =
       ]
       )
 
+let maybe_set_restart_for_all_vms ~__context ~repository_name ~updates_info =
+  let open Guidance in
+  let add_restart_to_vms ~__context ~vms ~kind =
+    let ( let* ) = Option.bind in
+    let* op =
+      match kind with
+      | Mandatory ->
+          Some Db.VM.add_pending_guidances
+      | Recommended ->
+          Some Db.VM.add_pending_guidances_recommended
+      | Full ->
+          Some Db.VM.add_pending_guidances_full
+      | _ ->
+          None
+    in
+    Some
+      (List.iter
+         (fun self ->
+           debug "add RestartVM for VM %s" (Ref.string_of self) ;
+           op ~__context ~self ~value:`restart_vm
+         )
+         vms
+      )
+  in
+  let vms =
+    Db.VM.get_all ~__context
+    |> List.filter (fun self ->
+           not (Db.VM.get_is_control_domain ~__context ~self)
+       )
+  in
+  let hosts = Db.Host.get_all ~__context in
+  if Hashtbl.length updates_in_cache = 0 then
+    set_available_updates ~__context |> ignore ;
+  let updates_of_hosts, _ =
+    consolidate_updates_of_hosts ~repository_name ~updates_info ~hosts
+  in
+  updates_of_hosts
+  (* fold each guidance kind from all hosts *)
+  |> List.fold_left
+       (fun acc HostUpdates.{guidance; _} ->
+         List.fold_left
+           (fun acc' (kind, l) ->
+             if List.mem RestartVM l && not (List.mem kind acc') then
+               kind :: acc'
+             else
+               acc'
+           )
+           acc guidance
+       )
+       []
```

Comment:
It's a little complicated for readability with using two folds. Here is a example using one fold:
```ocaml
updates_of_hosts
|> List.concat_map (fun HostUpdates.{guidance; _} -> guidance)
|> List.fold_left
  (fun acc (kind, l) ->
    if List.mem RestartVM l && not (List.mem kind acc) then
      kind :: acc
    else
      acc
  )
  []
```

Reply:
If ordering need to be maintained using `List.filter_map` is an option as well